src/chromatix/functional/propagation.py

```diff
@@ -274,32 +253,123 @@ def asm_propagate(
             Space Propagation in Far and Near Fields" (2009) by Matsushima and
             Shimobaba. Defaults to ``False``.
         shift_yx: If provided, defines a shift in microns in the destination
             plane. Should be an array of shape `[2,]` in the format `[y, x]`.
+        dx: If provided, defines a different output sampling at the output
+            plane.
+        N_out: If provided, defines the output shape of the field. Should be
+            a tuple of integers. If not provided and ``dx`` is provided, the
+            output shape will default to that of the input field.
+        use_czt: Whether or not to use chirp Z-transform for off-axis
+            propagation or interpolating with different output sampling.
+            Defaults to True if ``dx`` or ``N_out`` is provided.
         mode: Either "full" or "same". If "same", the shape of the output
             ``Field`` will match the shape of the incoming ``Field``. Defaults
             to "full", in which case the output shape will include padding.
     """
     field = pad(field, N_pad, cval=cval)
-    propagator = compute_asm_propagator(field, z, n, kykx, bandlimit, shift_yx)
-    field = kernel_propagate(field, propagator)
+    if dx is None and N_out is None:
+        # If neither dx nor N_out is provided, use the default ASM propagation
+        # as FFT is faster than CZT
+        use_czt = False
+    propagator = compute_asm_propagator(
+        field,
+        z,
+        n,
+        kykx,
+        bandlimit,
+        shift_yx if not use_czt else (0.0, 0.0),
+        remove_evanescent=remove_evanescent,
+    )
+    field = kernel_propagate(
+        field, propagator, dx=dx, N_out=N_out, shift_yx=shift_yx, use_czt=use_czt
+    )
     if mode == "same":
         field = crop(field, N_pad)
     return field
 
 
-def kernel_propagate(field: Field, propagator: Array) -> Field:
+def kernel_propagate(
+    field: Field,
+    propagator: Array,
+    dx: Union[float, Array] = None,
+    N_out: Tuple[int, int] = None,
+    shift_yx: Union[Array, Tuple[float, float]] = (0.0, 0.0),
+    use_czt: bool = False,
+) -> Field:
     """
     Propagate an incoming ``Field`` by the given propagation kernel
     (``propagator``). This amounts to performing a Fourier convolution of the
     ``field`` and the ``propagator``.
     """
     axes = field.spatial_dims
-    u = ifft(fft(field.u, axes=axes) * propagator, axes=axes)
+    if dx is None and N_out is None and not use_czt:
+        u = ifft(fft(field.u, axes=axes) * propagator, axes=axes)
+    else:
+        if N_out is None:
+            N_out = field.spatial_shape
+        if dx is None:
+            dx = field.dx.squeeze()
+        # TODO support vector field?
+        output_field = ScalarField.create(
+            dx=dx,
+            spectrum=field.spectrum,
+            spectral_density=field.spectral_density,
+            shape=N_out,
+            shift_yx=np.array(shift_yx)[..., np.newaxis],
```

Comment:
The shift gets baked at compile time? Is that what we want?

Reply:
It also doesn't add the existing shift of the incoming field. I expected that this is what we would want for the shift tracking. Is that not how this works?

Reply:
you're right, I was used to an API (in my previous code) where I just return the field, but we should also shift the grid in the input field. I'll come up with a utility for this

Reply:
Handled with the `shift_grid` method inside `src/chromatix/field.py`